### ISSUE TITLE
Run JSDoc programatically 

### ISSUE THREAD

**Issue by esatterwhite** [jsdoc3/jsdoc#938]
I would like to use jsdoc as a module rather than installing it globally and using the cli or forking a sub process, but it isn't very clear on how to do that.

Is this possible? How?

**Comment by hegemonic** [comment]
At this time, it's not really possible. I very much want to change this, but it will take a lot of effort to make JSDoc work programmatically. See #812 for some background.

For now, one viable option is to use [mrjoelkemp/jsdoc3-parser](https://github.com/mrjoelkemp/jsdoc3-parser), a clever hack that provides a programmatic wrapper around the JSDoc CLI tool.

This request comes up a _lot_, so I'm just going to keep this issue open, even though it's unlikely to be resolved anytime soon.

**Comment by esatterwhite** [comment]
It seems like a lot of the weirdness / complexity comes from trying to support Rhino. Is there really a need to keep support for Rhino any longer?

**Comment by hegemonic** [comment]
At this point, I'd say that almost none of the complexity of exposing JSDoc programmatically is related to Rhino support. Most parts of JSDoc don't know anything about Rhino and don't need to.

**Comment by esatterwhite** [comment]
I wasn't referring to the jsdoc parsing bits specifically. Everything in **not** in `lib/jsdoc`, with exception to stuff in `utils` seems really suspect to problems between runtimes. It looks like a lot of the work around comes from the `runtime.js` stuff, and how the jsdoc.js passes arguments into it. That is where it starts.

And how you have to conditionally require things into the jsdoc.js file based on if it is loaded from Rhino. You also have to effectively maintain a fork / mirror of Node.js core or polyfil them into the source code so they can be loaded into rhino. And vice versa - There are some copies of the node JSDoc source modified in the Rhino src tree. 

And there is also a big warning in  `cli` that the mirrored / augmented copies of  `fs` / `path` can & probably will fail on Rhino and it's understanding of `this`. 

There is also the `requizzle` part in `jsdoc.js` which, in most cases, a convenience during development, but if I want to use parts of JSDoc programmatically in a different project, I have to require the command line entry point, which monkey patches `require`. That causes problems in my project if I do, and it causes problems in JSdoc if I don't.

I think I can correct the cli and `app` / `env`  stuff to be much more streamlined and usable as library. But there is a lot of hoop jumping to get stuff to work across rumtimes. I would say 20% - 30% of the code is to deal with Rhino. 

Is it being kept around for the sake of legacy support? What is the benefit to maintaining support for Java?

**Comment by hegemonic** [comment]
Regarding Rhino support, I'll quote my comment from #942:

> We support Rhino for a few reasons:

> 1. In the past, we required Rhino.
> 2. For some users, it may be more convenient to run JSDoc on the JDK. (I have no idea whether this is true, or how many users would prefer the JDK over Node.js.)

> At some point, I plan to survey the community and find out how many people, if any, rely upon Rhino support. If nobody needs it, I'll deprecate the Rhino version. But for now, at least, we need to continue to support Rhino.

Believe me, I'd be happy to get rid of all the Rhino-related code. But I'm not going to remove Rhino support until I'm confident that dropping Rhino will not screw people over.

Regarding your concerns about Requizzle:

> There is also the `requizzle` part in `jsdoc.js` which, in most cases, a convenience during development, but if I want to use parts of JSDoc programmatically in a different project, I have to require the command line entry point, which monkey patches `require`. That causes problems in my project if I do, and it causes problems in JSdoc if I don't.

Requizzle only monkey-patches `require` calls within a) JSDoc and b) code required by JSDoc. It should not monkey-patch any code that JSDoc does not use.

**Comment by mnaughto** [comment]
I made some progress on this a while back on my own fork here: https://github.com/mnaughto/jsdoc.

I remember a discussion back when I was working on it about how those changes couldn't be pulled in because plugins depended on the global `env` and `app` variables. I wonder if there would be a way to alias them to the `Runtime` instance for backwards compatibility?

**Comment by 75lb** [comment]
[jsdoc-api](https://github.com/jsdoc2md/jsdoc-api) gives direct, programmatic access to the raw jsdoc output. It has sync, async and streaming interfaces. You can input source code as a string, a set of file names or a stream.

**Comment by esatterwhite** [comment]
@75lb I was looking more for a way to just configure and run jsodc programmatically  more so than have to parse the out put.

**Comment by 75lb** [comment]
me too, but that option doesn't yet exist so...

**Comment by esatterwhite** [comment]
@75lb can jsdoc-api render out the static html site as well?

**Comment by 75lb** [comment]
yeah.. i added a synchronous method to do that.. can add Promise and streaming interfaces too if the demand is there.. 

see [jsdoc.renderSync([options])](https://github.com/jsdoc2md/jsdoc-api#module_jsdoc-api.renderSync).

**Comment by jcrben** [comment]
As far as Rhino support, I don't see how removing it would screw anyone over. They would just have to use an older version.

**Comment by donaldpipowitch** [comment]
It looks like https://github.com/mrjoelkemp/jsdoc3-parser is quite old and https://github.com/jsdoc2md/jsdoc-api throws for me with example given in the README (https://github.com/jsdoc2md/jsdoc-api/issues/5).

Is there a better way to use Jsdoc programmatically by now?

**Comment by 75lb** [comment]
sorry, it was because the README example contained a unicode forward-slash, instead of the regular ASCII one.. try this instead: 
```
jsdoc.explainSync({ source: '/** example doclet */ \n var example = true' })
```

**Comment by onury** [comment]
@donaldpipowitch you can try https://github.com/onury/jsdoc-x

**Comment by dendrochronology** [comment]
Along the lines of the discussion above, is anyone here aware of a way to make JSDoc class/type info available at runtime? I'm thinking Python docstrings on steroids. Like a Babel plugin that injects `static __doc__ = { jsdoc: 'output', goes: 'here' }` or something.

Maybe that's a terrible idea, but I've found myself copying and pasting more than I should, which is always a red flag.

**Comment by matyasf** [comment]
Any update on this? Is it still not possible?
